File: frontend/src/utils/validators.js
/**
 * Validaciones del lado del cliente
 * Coinciden con las validaciones del backend para mejor UX
 */

/**
 * Valida la fortaleza de una contraseña
 * Criterios: mínimo 8 caracteres, una mayúscula, una minúscula, un número
 * @param {string} password - Contraseña a validar
 * @returns {Object} - {isValid: boolean, errors: string[]}
 */
export const validatePasswordStrength = (password) => {
  const errors = [];

  if (password.length < 8) {
    errors.push('La contraseña debe tener mínimo 8 caracteres');
  }

  if (!/[A-Z]/.test(password)) {
    errors.push('La contraseña debe contener al menos una mayúscula');
  }

  if (!/[a-z]/.test(password)) {
    errors.push('La contraseña debe contener al menos una minúscula');
  }

  if (!/\d/.test(password)) {
    errors.push('La contraseña debe contener al menos un número');
  }

  return {
    isValid: errors.length === 0,
    errors,
  };
};

/**
 * Calcula el nivel de fortaleza de una contraseña
 * @param {string} password - Contraseña a evaluar
 * @returns {string} - 'debil', 'media', 'fuerte', 'muy_fuerte'
 */
export const calculatePasswordStrength = (password) => {
  let score = 0;

  // Longitud
  if (password.length >= 8) score += 1;
  if (password.length >= 12) score += 1;
  if (password.length >= 16) score += 1;

  // Caracteres
  if (/[a-z]/.test(password)) score += 1;
  if (/[A-Z]/.test(password)) score += 1;
  if (/\d/.test(password)) score += 1;
  if (/[!@#$%^&*(),.?":{}|<>]/.test(password)) score += 1;

  if (score <= 2) return 'debil';
  if (score <= 4) return 'media';
  if (score <= 6) return 'fuerte';
  return 'muy_fuerte';
};

/**
 * Obtiene el color asociado al nivel de fortaleza
 * @param {string} strength - Nivel de fortaleza
 * @returns {string} - Color en formato CSS
 */
export const getPasswordStrengthColor = (strength) => {
  const colors = {
    debil: '#f44336', // Rojo
    media: '#ff9800', // Naranja
    fuerte: '#2196f3', // Azul
    muy_fuerte: '#4caf50', // Verde
  };
  return colors[strength] || '#9e9e9e';
};

/**
 * Obtiene el texto de label para el nivel de fortaleza
 * @param {string} strength - Nivel de fortaleza
 * @returns {string} - Texto descriptivo
 */
export const getPasswordStrengthLabel = (strength) => {
  const labels = {
    debil: 'Débil',
    media: 'Media',
    fuerte: 'Fuerte',
    muy_fuerte: 'Muy Fuerte',
  };
  return labels[strength] || '';
};

/**
 * Valida formato de email
 * @param {string} email - Email a validar
 * @returns {boolean} - True si el formato es válido
 */
export const validateEmailFormat = (email) => {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
};

/**
 * Valida que un valor no esté vacío
 * @param {string} value - Valor a validar
 * @returns {boolean} - True si no está vacío
 */
export const validateRequired = (value) => {
  return value && value.trim() !== '';
};

File: backend/app/utils/validators.py
import re
from email_validator import validate_email as email_lib_validate, EmailNotValidError


def validate_password_strength(password):
    """
    Valida la fortaleza de una contraseña según los criterios de CA-2:
    - Mínimo 8 caracteres
    - Al menos una mayúscula
    - Al menos una minúscula
    - Al menos un número

    Args:
        password: Contraseña a validar

    Returns:
        tuple: (is_valid: bool, errors: list)
    """
    errors = []

    if len(password) < 8:
        errors.append('La contraseña debe tener mínimo 8 caracteres')

    if not re.search(r'[A-Z]', password):
        errors.append('La contraseña debe contener al menos una mayúscula')

    if not re.search(r'[a-z]', password):
        errors.append('La contraseña debe contener al menos una minúscula')

    if not re.search(r'\d', password):
        errors.append('La contraseña debe contener al menos un número')

    is_valid = len(errors) == 0
    return is_valid, errors


def calculate_password_strength(password):
    """
    Calcula el nivel de fortaleza de una contraseña

    Args:
        password: Contraseña a evaluar

    Returns:
        str: 'debil', 'media', 'fuerte', 'muy_fuerte'
    """
    score = 0

    # Longitud
    if len(password) >= 8:
        score += 1
    if len(password) >= 12:
        score += 1
    if len(password) >= 16:
        score += 1

    # Caracteres
    if re.search(r'[a-z]', password):
        score += 1
    if re.search(r'[A-Z]', password):
        score += 1
    if re.search(r'\d', password):
        score += 1
    if re.search(r'[!@#$%^&*(),.?":{}|<>]', password):
        score += 1

    if score <= 2:
        return 'debil'
    elif score <= 4:
        return 'media'
    elif score <= 6:
        return 'fuerte'
    else:
        return 'muy_fuerte'


def validate_email_format(email, check_deliverability=True):
    """
    Valida el formato de un email usando email-validator

    Args:
        email: Email a validar
        check_deliverability: Si False, solo valida el formato sin verificar DNS

    Returns:
        tuple: (is_valid: bool, normalized_email: str or None, error: str or None)
    """
    try:
        validation = email_lib_validate(email, check_deliverability=check_deliverability)
        return True, validation.normalized, None
    except EmailNotValidError as e:
        return False, None, str(e)


def validate_role(role):
    """
    Valida que el rol sea uno de los permitidos

    Args:
        role: Rol a validar

    Returns:
        bool: True si el rol es válido
    """
    valid_roles = ['Admin', 'Gerente de Almacén', 'Personal de Ventas']
    return role in valid_roles
